### ISSUE TITLE
Rule RSPEC-907 reports for goto's used in switch statements to provide a state machine.

### ISSUE THREAD

**Issue by AraHaan** [SonarSource/sonar-dotnet#3996]
### Description

S907 is emitted from my zlib library which uses state machines as per the spec on deflate/inflate on zlib (https://github.com/Elskom/zlib.managed)

### Repro steps

first: the project https://github.com/Elskom/zlib.managed/ depends on things from the parent project which is https://github.com/Elskom/Sdk so clone that recursively.
Next open the solution inside of the root of the parent project then open up zlib.managed, 3 files use state machines, InfCodes.cs, InfBlocks.cs/InfTrees.cs, and Inflate.cs.

### Expected behavior

For no violations of the rule since removing ``goto`` on state machines breaks the entire state machine and in turn breaks how actually works.

### Actual behavior

The rule reports goto's in c# state machines as a violation. And trying to fix it then results in the state machine getting broken.

### Known workarounds

Currently the only way I could think of is to suppress the issue for now.

### Related information

* C#/VB.NET Plugins version : ??? I do not think I use the plugin with the c# analyzer package from nuget brought in via ``PackageReference``.
* Visual Studio version: VS2019 latest preview version
* MSBuild / dotnet version .NET SDK 5.0.102
* SonarScanner for .NET version (if used): The latest nuget package for the C# analyzer.
* Operating System
Windows 10 Latest Insider Preview (fast ring)

**Comment by AraHaan** [comment]
cc @ehartmann.

**Comment by pavel-mikula-sonarsource** [comment]
Hi @AraHaan ,

Can you please follow our guideline to report issues and provide minimal reproducible snippet?

>### Repro steps
>
>Please provide a code snippet that compiles which reproduces the problem. Alternatively, a link to a SonarCloud project where the issue appears, or a git repository, or attach a zip with a reproducer project.

Thank you

**Comment by pavel-mikula-sonarsource** [comment]
I'll fix the instructions to be clear about the need for minimal reproducer project in form of repo or zip.

**Comment by AraHaan** [comment]
```cs
internal sealed class InfCodes
{
    // waiting for "i:"=input,
    //             "o:"=output,
    //             "x:"=nothing
    private const int START = 0; // x: set up for LEN
    private const int LEN = 1; // i: get length/literal/eob next
    private const int LENEXT = 2; // i: getting length extra (have base)
    private const int DIST = 3; // i: get distance next
    private const int DISTEXT = 4; // i: getting distance extra
    private const int COPY = 5; // o: copying bytes in window, waiting for space
    private const int LIT = 6; // o: got literal, waiting for output space
    private const int WASH = 7; // o: got eob, possibly still output waiting
    private const int END = 8; // x: got eob and all data flushed
    private const int BADCODE = 9; // x: got error

    // other stuff
    // ctors that set the mode to START and the compression / decompression trees.
    
    internal int Mode { get; private set; } // current inflate_codes mode

    // other properties dealing with the compression / decompression trees made in the ctors.

    // some methods unreleated.

    internal ZlibCompressionState Proc(InfBlocks s, ZStream z, ZlibCompressionState r)
    {
        // create local variables and set them to initial values.
        // process input and output based on current state
        while (true)
        {
            switch (this.Mode)
            {
                // waiting for "i:"=input, "o:"=output, "x:"=nothing
                case START: // x: set up for LEN
                {
                    // do some work.
                    this.Mode = LEN;
                    goto case LEN;
                }

                case LEN: // i: get length/literal/eob next
                {
                    // do some work; return (with BADCODE) or break the switch with (LIT, LENEXT, or WASH) depending on some conditions.
                }

                case LENEXT: // i: getting length extra (have base)
                {
                    // do some work; return if certain conditions are true; else go to next switch case.
                    this.Mode = DIST;
                    goto case DIST;
                }

                case DIST: // i: get distance next
                {
                    // do some work; return or break into DISTEXT.
                }

                case DISTEXT: // i: getting distance extra
                {
                    // do some work.
                    this.Mode = COPY;
                    goto case COPY;
                }


                case COPY: // o: copying bytes in window, waiting for space
                {
                    // do some work; return or break into START based on conditional stuff.
                }

                case LIT: // o: got literal, waiting for output space
                {
                    // do some work; return or break into START (this.Mode = START followed by break) based on conditional stuff.
                }

                case WASH: // o: got eob, possibly more output
                {
                    // do some work; return if certain conditions are true; else go to the END case.
                    this.Mode = END;
                    goto case END;
                }


                case END:
                {
                    // return with END state code.
                }

                case BADCODE: // x: got error
                {
                    // return with BADCODE state code.
                }

                default:
                {
                    // return with zlib stream error code.
                }
            }
        }
    }
}
```

**Comment by pavel-mikula-sonarsource** [comment]
Hi @AraHaan,

Thank you for the code snippet. This `goto` usage is basically the problem that the rule is trying to prevent from. You have a state machine, that's not driven by it's state but manual jumps. And these jumps do not ensure that they jump to appropriate state. You can have something like this as a copy/paste mistake:
```
this.Mode = LEN;
goto case DIST;
```
If you replace all your `goto` statements with `break;` the `while(true)` will take care of the state and the state machine will be driven by it's state without a risk of error.
